File: src/brs/web/api/http/handler/SellAlias.java
package brs.web.api.http.handler;

import brs.*;
import brs.services.ParameterService;
import brs.util.Convert;
import brs.web.api.http.common.APITransactionManager;
import brs.web.api.http.common.LegacyDocTag;
import brs.web.api.http.common.ParameterException;
import com.google.gson.JsonElement;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;

public final class SellAlias extends CreateTransaction {

  private final ParameterService parameterService;
  private final Blockchain blockchain;

  public SellAlias(ParameterService parameterService, Blockchain blockchain, APITransactionManager apiTransactionManager) {
    super(new LegacyDocTag[] {LegacyDocTag.ALIASES, LegacyDocTag.CREATE_TRANSACTION}, apiTransactionManager, ALIAS_PARAMETER, ALIAS_NAME_PARAMETER, TLD_PARAMETER, RECIPIENT_PARAMETER, PRICE_NQT_PARAMETER);
    this.parameterService = parameterService;
    this.blockchain = blockchain;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) throws SignumException {
    Alias alias = parameterService.getAlias(req);
    Account owner = parameterService.getSenderAccount(req);

    String priceValueNQT = Convert.emptyToNull(req.getParameter(PRICE_NQT_PARAMETER));
    if (priceValueNQT == null) {
      return MISSING_PRICE;
    }
    long priceNQT;
    try {
      priceNQT = Long.parseLong(priceValueNQT);
    } catch (RuntimeException e) {
      return INCORRECT_PRICE;
    }
    if (priceNQT < 0 || priceNQT > Constants.MAX_BALANCE_NQT) {
      throw new ParameterException(INCORRECT_PRICE);
    }

    String recipientValue = Convert.emptyToNull(req.getParameter(RECIPIENT_PARAMETER));
    long recipientId = 0;
    if (recipientValue != null) {
      try {
        recipientId = Convert.parseAccountId(recipientValue);
      } catch (RuntimeException e) {
        return INCORRECT_RECIPIENT;
      }
      if (recipientId == 0) {
        return INCORRECT_RECIPIENT;
      }
    }

    if (alias.getAccountId() != owner.getId()) {
      return INCORRECT_ALIAS_OWNER;
    }

    Attachment attachment = new Attachment.MessagingAliasSell(alias.getId(), alias.getAliasName(), priceNQT, blockchain.getHeight());
    return createTransaction(req, owner, recipientId, 0, attachment);
  }
}

File: src/main/kotlin/brs/http/SellAlias.kt
package brs.http

import brs.Attachment
import brs.Constants
import brs.DependencyProvider
import brs.http.JSONResponses.INCORRECT_ALIAS_OWNER
import brs.http.JSONResponses.INCORRECT_PRICE
import brs.http.JSONResponses.INCORRECT_RECIPIENT
import brs.http.JSONResponses.MISSING_PRICE
import brs.http.common.Parameters.ALIAS_NAME_PARAMETER
import brs.http.common.Parameters.ALIAS_PARAMETER
import brs.http.common.Parameters.PRICE_PLANCK_PARAMETER
import brs.http.common.Parameters.RECIPIENT_PARAMETER
import brs.util.convert.emptyToNull
import brs.util.convert.parseAccountId
import com.google.gson.JsonElement
import javax.servlet.http.HttpServletRequest

internal class SellAlias internal constructor(private val dp: DependencyProvider) : CreateTransaction(dp, arrayOf(APITag.ALIASES, APITag.CREATE_TRANSACTION), ALIAS_PARAMETER, ALIAS_NAME_PARAMETER, RECIPIENT_PARAMETER, PRICE_PLANCK_PARAMETER) {

    override fun processRequest(request: HttpServletRequest): JsonElement {
        val alias = dp.parameterService.getAlias(request)
        val owner = dp.parameterService.getSenderAccount(request)

        val priceValuePlanck = request.getParameter(PRICE_PLANCK_PARAMETER).emptyToNull() ?: return MISSING_PRICE
        val pricePlanck: Long
        try {
            pricePlanck = java.lang.Long.parseLong(priceValuePlanck)
        } catch (e: RuntimeException) {
            return INCORRECT_PRICE
        }

        if (pricePlanck < 0 || pricePlanck > Constants.MAX_BALANCE_PLANCK) {
            throw ParameterException(INCORRECT_PRICE)
        }

        val recipientValue = request.getParameter(RECIPIENT_PARAMETER).emptyToNull()
        var recipientId: Long = 0
        if (recipientValue != null) {
            try {
                recipientId = recipientValue.parseAccountId()
            } catch (e: RuntimeException) {
                return INCORRECT_RECIPIENT
            }

            if (recipientId == 0L) {
                return INCORRECT_RECIPIENT
            }
        }

        if (alias.accountId != owner.id) {
            return INCORRECT_ALIAS_OWNER
        }

        val attachment = Attachment.MessagingAliasSell(dp, alias.aliasName, pricePlanck, dp.blockchain.height)
        return createTransaction(request, owner, recipientId, 0, attachment)
    }
}
